### DISCUSSION TITLE
“Atomic” timing of IRQs

### DISCUSSION THREAD

**Discussion by mdaeron** [micropython/micropython#10866] [General]
I am wondering what kind of “atomicity” there is in the treatment of regular MicroPython instructions and interrupt callbacks. Let's say I want to print out or modify the contents of an array:

```py
for k in range(len(myarray)):
    myarray[k] = myarray[k] * 2
print(myarray)
```

Is there any chance that an IRQ will modify the contents of `myarray` while in the `for` loop (I believe the answer is yes)? What about during the `print()` call?

What are the best practices when manipulating arrays which might potentially be modified by IRQs? Should I make a copy of such an array first, as in: `myarray_copy[:] = myarray[:]`, presuming that such a single line of code will be treated as “atomic” (non-interruptable)?

**Comment by robert-hh** [comment]
You can disable and enable interrupts during critical phases of your code. See https://docs.micropython.org/en/latest/library/machine.html?highlight=disable#machine.disable_irq

**Comment by peterhinch** [comment]
These docs discuss the issues involved with concurrency. 
[Official docs](http://docs.micropython.org/en/latest/reference/isr_rules.html)  
[Threading](https://github.com/peterhinch/micropython-async/blob/master/v3/docs/THREADING.md) includes details of ISR concurrency.  
[Alternatives to interrupts](https://github.com/peterhinch/micropython-async/blob/master/v3/docs/INTERRUPTS.md) Does an application really need an IRQ?
